CSHARP-2405: ImmutableTypeClassMapConvention should not consider static properties

sindbach opened the pull request:
-Included Ross Buggins PR 
-Added Ross as a contributor in README
-Added test

Evergreen: https://evergreen.mongodb.com/version/5da518e0e3c33150766585aa

rbugginsvia committed:
Update ImmutableTypeClassMapConvention.cs

Upate to fix problem with static members in F# not allowing serialisation
https://jira.mongodb.org/browse/CSHARP-2405

The above limits the GetProperties to just instance public properties - ignoring static properties.

sindbach committed:
CSHARP-2405: ImmutableTypeClassMapConvention should not consider static properties

rstam commented on tests/MongoDB.Bson.Tests/Serialization/Conventions/ImmutableTypeClassMapConventionTests.cs:
```diff
@@ -60,6 +60,15 @@ public TestClassC(string a, string b)
         }
 
+        private class TestClassD
+                {
```
Fix indentation

rstam reviewed (approved):
Fix minor indentation issue and then LGTM.

sindbach committed:
CSHARP-2405: ImmutableTypeClassMapConvention should not consider static properties

DmitryLukyanov commented on src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs:
```diff
@@ -39,7 +39,7 @@ public void Apply(BsonClassMap classMap)
             }
 
-            var properties = typeInfo.GetProperties();
+            var properties = typeInfo.GetProperties(BindingFlags.Instance | BindingFlags.Public );
```
this change affects two code paths:
1. https://github.com/sindbach/mongo-csharp-driver/pull/3/files#diff-1bcd6b847860a1520a8a7ea2b958170aR44
2. https://github.com/sindbach/mongo-csharp-driver/pull/3/files#diff-1bcd6b847860a1520a8a7ea2b958170aR53

But the created test validated only the second one. Should we add a test to check the first?

DmitryLukyanov reviewed (approved):
LGTM, with small comments:
1. Considering that these formatting issues are fixed:
https://github.com/sindbach/mongo-csharp-driver/pull/3/files#diff-c8c099d3b54a1bfbf3b68bbf10cd7887R67
https://github.com/sindbach/mongo-csharp-driver/pull/3/files#diff-c8c099d3b54a1bfbf3b68bbf10cd7887R147
2. Suggested to add a new test which will validate a static field with `CanWrite = true`

rstam reviewed (changes requested):
@sindbach go ahead and fix the formatting issues and add the test suggested by @DmitryLukyanov and re-request our review.

DmitryLukyanov commented:
NOTE: I cannot open linked evergreen patch

sindbach committed:
CSHARP-2405: ImmutableTypeClassMapConvention should not consider static properties

sindbach commented:
@DmitryLukyanov Thanks for the review. I've fixed the formatting (the extra space and the expression body for properties) and added a new test for a class with a writeable property.  

The new evergreen patch is : https://evergreen.mongodb.com/version/5dc1117861837d7f5d6e2755 
Also, in regards to the access for you, it's likely under the internal evergreen ? 

Cheers.

DmitryLukyanov commented on tests/MongoDB.Bson.Tests/Serialization/Conventions/ImmutableTypeClassMapConventionTests.cs:
```diff
@@ -125,5 +143,25 @@ public void TestNoDefaultConstructorClassMapConventionWithTestClassC()
+        }
+
+        [Fact]
+        public void TestNoDefaultConstructorClassMapConventionWithTestClassE()
```
this test is ok. 
But since the change from this ticket excluded the static properties. I expected the test for this class:

        private class TestClassF
        {
            public TestClassD(int a) { A = a; }

            public static TestClassD Default { get; set; } = new TestClassD(0);

            public int A { get; }
        }

So, I suggest adding one more test.

DmitryLukyanov commented:
> Also, in regards to the access for you, it's likely under the internal evergreen ?

yesterday we had an issue with evergeen. Now, I have access to your patch.

sindbach committed:
CSHARP-2405: ImmutableTypeClassMapConvention should not consider static properties

sindbach commented:
New test class added. 
Evergreen: https://evergreen.mongodb.com/version/5dc24f5e0ae60653b21dd116

DmitryLukyanov reviewed (approved):
LGTM

rstam reviewed (approved):
LGTM

vincentkam reviewed (approved):
LGTM! 😸 Thanks @sindbach!

vincentkam commented:
Since we are trying to get 2.9.3 out today, I've merged this PR upstream.
